### PR TITLE
Add TypeScript type declarations

### PR DESCRIPTION
**Pull request by jurca**
This PR adds TypeScript type declarations for all implementation files, improving interoperability with the TypeScript ecosystem.

Please consider accepting this PR in favor of setting up a separate @types/ package for this.

### CODE REVIEW

**Comment by ljharb** [comment]
Thanks, but I'm not interested in TypeScript at this time. Please create a DT package instead.

**Comment by jurca** [comment]
:disappointed: 

OK, I understand. Thanks for the reply.

File: package.json [hunk 1]
```diff
@@ -17,7 +22,7 @@
 		"pretest": "npm run --silent lint && es-shim-api --bound",
 		"test": "npm run --silent tests-only",
 		"posttest": "npx aud",
-		"tests-only": "npm run --silent test:shimmed && npm run --silent test:module && npm run --silent tests:es5",
+		"tests-only": "npm run --silent test:shimmed && npm run --silent test:module && npm run --silent tests:es5 && npx tsc",
```
Comment (ljharb) [inline package.json:25]: does this not need `-p` to work?

File: implementation.d.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+interface PromiseRejection<E extends Error = Error> {
+	status: 'rejected';
+	reason: E;
+}
+
+type PromiseResult<T, E extends Error = Error> = PromiseResolution<T> | PromiseRejection<E>;
+
+type PromiseTuple1<T1> = [Promise<T1>];
```
Comment (ljharb) [inline implementation.d.ts]: is writing out all the tuples really the best way typescript has to handle this?

File: implementation.d.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+interface PromiseResolution<T> {
+	status: 'fulfilled';
+	value: T;
+}
+
+interface PromiseRejection<E extends Error = Error> {
```
Comment (ljharb) [inline implementation.d.ts]: Promises need not reject with an error; this should just be E.

File: implementation.d.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	status: 'rejected';
+	reason: E;
+}
+
+type PromiseResult<T, E extends Error = Error> = PromiseResolution<T> | PromiseRejection<E>;
```
Comment (ljharb) [inline implementation.d.ts]: same here

File: package.json (hunk 1 above)
Comment (jurca) [reply package.json:25]: Interestingly enough, no. I assume this is one of the more recent changes in TypeScript.

File: implementation.d.ts (hunk 2 above)
Comment (jurca) [reply implementation.d.ts]: That's a matter of opinion, I would say. It's one possible way, I do not insist on writing the code this way. The tuple types could be inlined in the function declaration and it would work just fine.

File: implementation.d.ts (hunk 3 above)
Comment (jurca) [reply implementation.d.ts]: Technically true, but a bad practice since you will get no information on where the error originated from due to lack of a stack trace.

I would argue going the stricter way, since it is common with TypeScript. While the following code works in JS:

```javascript
parseInt(10, 10)
```

You would get an error in TypeScript:

```
error TS2345: Argument of type '10' is not assignable to parameter of type 'string'.
parseInt(10, 10)
         ~~
```

On the other hand, TypeScript does allow you to pass non-errors to the `throw` operator.

I'll leave the final decision on you.

File: implementation.d.ts (hunk 2 above)
Comment (ljharb) [reply implementation.d.ts]: i mean, is there no way to avoid any of them, and just say “all the types passed in, map to these passed out”?

File: implementation.d.ts (hunk 3 above)
Comment (ljharb) [reply implementation.d.ts]: Whatever throw allows is what a promise rejection allows; it’s a type bug if you pass a number to parseInt but not if you reject with `false`, say.

Comment (jurca) [reply implementation.d.ts]: Very well.

File: implementation.d.ts (hunk 2 above)
Comment (jurca) [reply implementation.d.ts]: After more research, I was able to find this: https://github.com/Microsoft/TypeScript/pull/26063

To be honest, I was not able to find a solution to this previously, so I went with an approach I saw in the reselect library.

I will fix this, thanks for bringing this up.

**Comment by jurca** [comment]
@ljharb I have updated the type definitions with the agreed-upon changes, sorry for the delay. I have also dropped the type variables for error types, since it's impossible for typescript to guarantee them.

File: implementation.d.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+	value: T;
+}
+
+interface PromiseRejection<E> {
+	status: 'rejected';
+	reason: E;
+}
+
+type PromiseResult<T, E> = PromiseResolution<T> | PromiseRejection<E>;
+type PromiseTuple<T extends [any, ...any[]]> = {[P in keyof T]: Promise<T[P]>};
```
Comment (ljharb) [inline implementation.d.ts]: hmm, does this need `any` here, or can it use `unknown`?

**Comment by ljharb** [comment]
It's impossible using the spread form, you mean, or at all?

File: implementation.d.ts (hunk 5 above)
Comment (jurca) [reply implementation.d.ts]: Since the library is not implemented in TypeScript, it makes no difference here IMHO. The client code specifies the actual types using type variables/inference.

**Comment by jurca** [comment]
> It's impossible using the spread form, you mean, or at all?

At all. There have been various suggestions to implement some form of type checking for thrown errors, but none of those I seen got accepted.

I originally introduced them for possible future support for this, but given the cost, effort and performance impact of such a feature, I find it somewhat unlikely to be implemented.

The spread form would require specifying somehow that both type enums have the same length, which is possible, just a little bit inconvenient, for example:

```typescript
type PromiseResultTuple<
    T extends B,
    E extends B,
    B extends [any, ...any[]],
> = {[P in keyof B]: PromiseResult<T[P], E[P]>}
```

Another approach:
```typescript
type PromiseResultTuple<
    T extends B,
    E extends B,
    L extends number,
    B extends [any, ...any[]] & { length: L } = [any, ...any[]] & { length: L },
> = { [P in keyof B]: PromiseResult<T[P], E[P]> }
```

**Comment by ljharb** [comment]
I'm not experienced enough with TS to know what's better, but it seems that any use of `any` is terrible, and any loss of type information as things pass through is a huge shame.

**Comment by jurca** [comment]
In general, when writing TypeScript code, `any` should be avoided in favor of `unknown` except for a few cases. Since these are type declarations and your JS code is not checked using them, it makes little difference - except for one spot: the type of thrown value in a rejected promise result.

So, I changed all `any`s to `unknown`s, but the effect is pretty much cosmetic.

**Comment by ljharb** [comment]
I've taken another look at this - `npx tsc` will only verify that the types are self-consistent, but it won't verify that they actually match the API. It seems important to me to ensure that they don't drift over time.

**Comment by jurca** [comment]
I like that you are approaching this issue so responsibly, however, resolving this will not be so easy.

What you noticed about tsc is, unfortunately, an issue with all DefinitelyTyped packages, but the community seems to be able to keep up pretty well.

It is - probably - better when the type definitions are maintained by the package author as a part of the package itself, but, as you pointed out, there is no reliable way to ensure that the type definitions are up to date (at least I cannot think of any that would be more convenient to maintain than the option below).

The most reliable future-proof way would be to port the project to TypeScript and have the type definitions generated automatically by tsc. I am willing to do this and promise to help maintaining the project in the future, however, you are not interested in TypeScript (and I respect that), so I am not sure how to move on from here.

**Comment by ljharb** [comment]
I'd be happy writing in JS, and manually maintaining the type defs in `.d.ts` files, as long as i had some tooling that could remind me when I needed to update them. The worst case would be that for every release, i have to publish an additional patch just because i forgot to update the types.

It seems like there could be a tool (someone might need to make it) that temporarily renames the js files to `.ts` files, generates the types, and compares the generated d.ts files? Would you be willing to try to write that tool?

Once it exists, I'd be willing to start shipping first-class types alongside many of my JS packages.

**Comment by jurca** [comment]
I have taken some time to look into this, exploring various options.

TL;DR: Such tool does not exist, and it cannot be with reliable results. There is another option, see below.

Long story:
First I explored enabling JS type checking, but that does not allow generating type definitions:
`tsconfig.json:9:3 - error TS5053: Option 'allowJs' cannot be specified with option 'declaration'.`

Since I would still need to use TypeScript's parser to properly read the contents of said declarations to validate them, generating declarations would not be much of a help anyway.

Renaming `.js` files to `.ts` (and removing any strict type checking restrictions) would allow me to process these, but since the type information is not there (you cannot have `.d.ts` files for introducing types into `.ts` files), yielded type information to check would be very generic. For example, this is what the type information of `implementation.js` would look like:

```typescript
declare const _default: (iterable: any) => any;
export = _default;
```

Getting this vague information is not straightforward either - first, commonjs import/export syntax would have to be transpiled using custom code transformer in the renamed `.js->.ts` files - see https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require.

Even with this information, there is not much to check except whether the exports more-or-less match, and - as you can see - I cannot verify that the argument types or return types are valid because the information simply is not there. BTW - it turns out you are not the first to suggest this approach: https://stackoverflow.com/questions/44146105/can-typescript-verify-that-a-js-file-is-consistent-with-its-own-d-ts-declarati#answer-44150588.

So, what is this other option? Using [JSDoc comments](https://github.com/Microsoft/TypeScript/wiki/Type-Checking-JavaScript-Files#jsdoc-types-are-used-for-type-information) in `.js` files without renaming them, parsing them using TypeScript parser and exporting the extracted type information into `.d.ts` files (remember, `tsc` does not allow generating `.d.ts` files when processing `.js` files). This option, however, is not ideal, since the type definitions would have to be simplified due to [lack of signature overloading support in JSDoc comments](https://github.com/Microsoft/TypeScript/issues/25590), which would make the library less convenient to use. For example, this code:

```typescript
Promise.allSettled(Promise.resolve("a"), Promise.resolve(1))
```

would have a result of the following type:

```typescript
Promise<Array<string | number>> // no length check, every item is number | string
```

instead of:
```typescript
Promise<[string, number]> // checks length, first item is string, second item is number
```

Therefore, if you want high-quality type declarations for your code, and you want to be 100 % sure they are up to date, I am afraid there is no other option than to port the project into TypeScript.

**Comment by jurca** [comment]
Hi, do you have any additional thoughts on this?

Is the current form of type declarations good enough for you, so that I can publish them in @types, or do you want to take a different approach?

**Comment by ljharb** [comment]
Sorry for the delay. It’s totally fine to publish them in DT in the meantime; but I’m still playing with options to see how the type shape can be compared programmatically to the js shape.

**Comment by jurca** [comment]
Thanks, I'll publish them.

I hope you'll be more successful in finding a reliable solution to comparing type declarations to JS code than I was.

**Comment by jurca** [comment]
Also, thanks for looking into this so much and all the feedback you have provided me, it was most helpful! :+1:

**Comment by jurca** [comment]
Pull request for type definitions in DefinitelyTyped can be tracked [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36314).

**Comment by ljharb** [comment]
Closing, since the DT types are published.
